core/src/main/java/org/apache/calcite/runtime/SqlFunctions.java

```diff
@@ -4107,6 +4113,97 @@ public static List arrayExcept(List list1, List list2) {
     return new ArrayList<>(result);
   }
 
+  /** Support the ARRAY_INSERT(array, pos, val) function. */
+  public static @Nullable List arrayInsert(List baselist, Object pos, Object val) {
+    if (baselist == null || pos == null) {
+      return null;
+    }
+    int posInt = (int) pos;
+    if (posInt == 0) {
+      throw new IllegalArgumentException("The index 0 is invalid. "
+          + "An index shall be either < 0 or > 0 (the first element has index 1)");
+    }
+
+    boolean usePositivePos = posInt > 0;
+
+    if (usePositivePos) {
+      int newArrayLength = Math.max(baselist.size() + 1, posInt);
```

Comment:
and here posInt is positive, if posInt < 0, it will from left to insert. 

And the logic of inserting null is the same, as long as `pos` is greater than the length of the array, then calculate the new array length, and fill in the middle with `null`

Reply:
I get it. Thanks for your explanation. How about we use the `System.arraycopy()` and `Arrays.fill()` to implement this? Others are good to me.

Reply:
@NobiGo yes. you are right. `System.arraycopy()` may has better performance when array is large. I have fixed it. PTAL.